src/os_win32.c

```diff
@@ -7689,31 +7700,141 @@ mch_setenv(char *var, char *value, int x)
     return 0;
 }
 
-#ifndef FEAT_GUI_W32
-
 /*
  * Support for 256 colors and 24-bit colors was added in Windows 10
  * version 1703 (Creators update).
  */
-# define VTP_FIRST_SUPPORT_BUILD MAKE_VER(10, 0, 15063)
+#define VTP_FIRST_SUPPORT_BUILD MAKE_VER(10, 0, 15063)
+
+/*
+ * Support for pseudo-console (ConPTY) was added in windows 10
+ * version 1809 (October 2018 update).
+ */
+#define CONPTY_FIRST_SUPPORT_BUILD MAKE_VER(10, 0, 17763)
+
+#ifdef FEAT_GUI_W32
+static HWINSTA origsta;
+static HWINSTA newsta;
+static HDESK origdesk;
+static HDESK newdesk;
+static PROCESS_INFORMATION pista;
+#endif
+
+    static void
+vtp_flag_init(void)
+{
+    DWORD   ver = get_build_number();
+    DWORD   mode;
+    HANDLE  out;
+
+#ifdef FEAT_GUI_W32
+    char    name[256];
+    char    startup[256];
+    char_u  *cmd;
+    char    cmdline[256];
+    SECURITY_ATTRIBUTES sa;
+    STARTUPINFO		si;
+
+    if (ver >= CONPTY_FIRST_SUPPORT_BUILD)
+    {
+	/* A console opened on another window station, get its standard
+	 * output. */
+
+	vim_memset(&sa, 0, sizeof(sa));
+	sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+	sa.bInheritHandle = TRUE;
+
+	origdesk = GetThreadDesktop(GetCurrentThreadId());
+	origsta = GetProcessWindowStation();
+
+	newsta = CreateWindowStation(NULL, 0, WINSTA_ALL_ACCESS, &sa);
+
+	if (SetProcessWindowStation(newsta))
+	{
+	    newdesk = CreateDesktop("Default", NULL, NULL, 0, GENERIC_ALL,
+									  &sa);
+	    SetThreadDesktop(newdesk);
+	    name[0] = NUL;
+	    GetUserObjectInformation(newsta, UOI_NAME, name, sizeof(name),
+									 NULL);
+	    sprintf(startup, "%s\\Default", name);
```

Comment:
According to the logs I left behind.
The reason for using the new desktop is that CreatePseudoConsole() uses STD_OUTPUT_HANDLE which is only in the console window of the SW_SHOW state in the same process.

This is the best policy because SW_HIDE disables STD_OUTPUT_HANDLE.

The reason for using CONOUT$ is because it is a file handle without entities.

Reply:
Ah, I see. `CreatePseudoConsole` _only_ uses the handles you pass when you call it, and no STD_OUTPUT_HANDLE.
When you use CreateProcess for the _shell_, you just need to make sure it does not inherit Vim's STD_OUTPUT_HANDLE.
My team wrote that API :smile:

Reply:
It is an old memory so I made a mistake a bit.
STD_OUTPUT_HANDLE for using ENABLE_VIRTUAL_TERMINAL_PROCESSING does not exist in gVim which is the calling process of CreatePseudoConsole().
Formally, how should I manipulate the ENABLE_VIRTUAL_TERMINAL_PROCESSING flag from the SUBSYSTEM:WINDOWS application?

Reply:
You don't need to manipulate `ENABLE_VIRTUAL_TERMINAL_PROCESSING` for ConPTY. ConPTY translates all console API calls into VT for you.

It is bad to set `ENABLE_VIRTUAL_TERMINAL_PROCESSING` for random programs, because they might not be expecting it.

Reply:
I thought the main power supply of ConPTY was EVTP flag.
I understood that it is independent.
If so, I will delete this window station.
Thank you for your teachings.